Fix latency issues with the ping protocol

luckysori opened the pull request:
Fixes https://github.com/itchysats/itchysats/issues/2050.

Patch 206bc0a8d802f1c1683ea55475aa782288436b0e might be contentious (and does not need to be merged), but I am fairly sure that it is the way to go.

---

Check out https://github.com/itchysats/itchysats/commit/5882fd5851d3a36b886f651edd4a17b58154c44a to see how I tested it and read some observations.

luckysori committed:
Move and rename xtra_libp2p_ping::ping submodule

luckysori committed:
Set xtra_libp2p_ping::protocol functions as pub(crate)

luckysori committed:
Create xtra_libp2p_ping::ping submodule

luckysori committed:
Split xtra_libp2p_ping::Actor into two

Fixes https://github.com/itchysats/itchysats/issues/2050.

By splitting the original `xtra_libp2p_ping::Actor` into two simpler
actors we ensure that the handling of external messages (i.e. pongs
from the counterparty) is never halted by the processing of internal
messages.

Given how `xtra_libp2p::Endpoint` is designed, it is unnecessary (and
sometimes harmful) to combine the protocol roles of listener and
dialer in a single actor. Protocol listener actors such as `pong` are
registered with the `Endpoint` when it is constructed. Protocol
dialers (they don't need to be actors!) such as `ping::Actor` can
request a negotiated substream from the `Endpoint` to kick off (and
continue) the protocol.

luckysori committed:
Conclude ping task indepent of recording latency

luckysori committed:
Ensure that recording ping latency cannot panic

luckysori committed:
Require interval when configuring `ping::Actor`

Otherwise we can configure the actor to do nothing.

luckysori committed:
Retry until latency is returned in test

Rather than hard-coding an arbitrarily long sleep.

luckysori committed:
Separate fetching connected peers from pinging

We don't want to block sending out all pings on synchronising our
internal state with the `Endpoint`. This could delay pings that the
listener is expecting.

We may end up failing to send pings to disconnected peers a bit more
frequently, but this was already possible before as you could not rely
on listeners still being connected by the time we sent out the ping:
the connection could have been dropped in between.

luckysori committed:
Stop takers from pinging the maker

Running a ping protocol between takers and makers serves several
purposes:

- Much like the heartbeat, using the connection regularly should help
keep it alive.
- The maker can gather metrics on the health of the connection between
itself and the takers that are connected to it.

Before this patch, each taker was also sending pings to the maker.
This was redundant (maker is already keeping the connection alive) and
wasteful (the takers were not doing anything with this information).

It was theorised that takers could eventually use their latency to the
maker to decide whether they should attempt to reconnect. I think
there are alternative ways to achieve this. For example, the taker
could trigger a reconnect if he has not received a ping from the maker
in a long time.

klochowicz commented on daemon/src/lib.rs:
```diff
@@ -202,11 +201,7 @@ where
-            .create(None)
-            .spawn(&mut tasks);
         let pong_address = pong::Actor::default().create(None).spawn(&mut tasks);
-
```
I feel like this should be a separate PR, as you suspect that it's a contentious change - let's chat about it separately.
the main reason is: it's hard to observe the possible congestions you're mitigating in this PR  if you disable that side of the protocol. 

all in all, I'm happy to explore alternatives, but I don't think they'll be easier - the one outlined in the commit message has the same drawback that the old heartbeat (intimate knowledge how pings on the other party are set up, ie. how often do they ping). Pinging from the taker overcomes this. If anything, I'd consider removing the heartbeat from the *maker* side - we don't really do anything with the latency measurements.

Also, we're not using this information *yet*, because we still rely on heartbeat - hopefully not for a long time.

luckysori replied:
> I feel like this should be a separate PR, as you suspect that it's a contentious change - let's chat about it separately.
the main reason is: it's hard to observe the possible congestions you're mitigating in this PR if you disable that side of the protocol.

I am happy to move it. I was just lazy last night.

I just made a lot of observations using [the test](https://github.com/itchysats/itchysats/commit/5882fd5851d3a36b886f651edd4a17b58154c44a) and by far the biggest difference with the `1-to-m` actor setup was to not have the `m`s pinging too.

> all in all, I'm happy to explore alternatives, but I don't think they'll be easier - the one outlined in the commit message has the same drawback that the old heartbeat (intimate knowledge how pings on the other party are set up, ie. how often do they ping). Pinging from the taker overcomes this. If anything, I'd consider removing the heartbeat from the maker side - we don't really do anything with the latency measurements.

This is fair. We don't know what the best way to trigger reconnecting is yet. I still don't think pinging in both directions is wise, but I could be wrong.

klochowicz commented on daemon/src/lib.rs:
```diff
@@ -200,17 +202,18 @@ where
         let (endpoint_addr, endpoint_context) = Context::new(None);
 
-        let ping_address = xtra_libp2p_ping::Actor::new(endpoint_addr.clone(), PING_INTERVAL)
+        ping::Actor::new(endpoint_addr.clone(), PING_INTERVAL)
```
It's a nifty way of getting around of congestion in ping, but I must admit am not in favour of this commit  - it just hides the underlying problem (congestion on processing external messages and trying to handle internal ones).

I feel like we need a more general solution, as this cannot be used in any protocols that require more interaction (e.g. rollover). 

the ping protocol problems we've seen is only a manifestation of problems that we see in other places, and should not be treated specially. We're looking at it as it's more isolated and easier to reproduce these problems - we can't roll such an approach to cfd actors etc.

We'll likely need a new way of scheduling internal messages in an async way, or a 2-tier queue (internal messages taking priority over external, or something else entirely!).

 out of curiosity: have you tried disabling prometheus metrics for the tests? has that changed anything?

luckysori replied:
> It's a nifty way of getting around of congestion in ping, but I must admit am not in favour of this commit - it just hides the underlying problem (congestion on processing external messages and trying to handle internal ones).

I see no benefit in having one actor for this protocol. The split shows that the `ping` and `pong` actors share no code. They are doing different things. They don't need to share state (the `pong` actor is stateless). It makes no sense to block the processing of pings on processing messages that access internal state that we don't care about!

I think the original design was flawed because it unnecessarily entangled different concerns into a single actor. It may be attractive to put everything in the same actor, but we have to have a strong reason to do so. After all, an actor is just a mechanism to perform actions concurrently based on state.

> I feel like we need a more general solution, as this cannot be used in any protocols that require more interaction (e.g. rollover).
>
> the ping protocol problems we've seen is only a manifestation of problems that we see in other places, and should not be treated specially. We're looking at it as it's more isolated and easier to reproduce these problems - we can't roll such an approach to cfd actors etc.

My learning from this patch is that we should try to build smaller actors. I suspect that we can see performance improvements by applying this idea to other parts of the codebase.

> We'll likely need a new way of scheduling internal messages in an async way, or a 2-tier queue (internal messages taking priority over external, or something else entirely!).

I am absolutely for introducing extensions to `xtra`, either by contributing or building on top! I think there is potential to make performance improvements that way too.

> out of curiosity: have you tried disabling prometheus metrics for the tests? has that changed anything?

I did, but I saw no notable improvements :(

klochowicz replied:
> I think the original design was flawed because it unnecessarily entangled different concerns into a single actor. It may be attractive to put everything in the same actor, but we have to have a strong reason to do so. After all, an actor is just a mechanism to perform actions concurrently based on state.

an actor is an abstraction, and in this case I think the original design nailed the level of abstraction - abstracting one protocol per actor. IMHO this should not cause trouble if we didn't have some underlying problems. 

In recent work we started to avoid internal dispatch to actually achieve some concurrency - otherwise, if all the futures need to fan back in 1 by 1 to complete an internal dispatch, we love all the benefits! 

thanks for checking disabling the metrics, at least that tells us it's not the issue :-)

luckysori replied:
> an actor is an abstraction, and in this case I think the original design nailed the level of abstraction - abstracting one protocol per actor. IMHO this should not cause trouble if we didn't have some underlying problems.

Sorry but I think this is wrong.

What is the benefit of putting the dialer and listener protocols in the same actor (other than code organisation, which is subjective), at the cost of fundamentally limiting our ability to perform the listener protocol in a timely manner?

By combining both protocols we are using one message queue when we need two.

Even the design of `xtra_libp2p` guides us towards only registering the listener protocol actor when constructing `Endpoint`: the dialer protocol works without passing its `xtra::Address` to the constructor. And the dialer protocol doesn't even need to be managed by an actor!

klochowicz reviewed (changes requested):
I love the refactor and a few small fixups, but I do not support the 2 congestial changesets:
- splitting the actor (this just hides the problem),
- removing the ping from the taker

Any chance we can move them out of this PR so that we can discuss this is stages?

da-kami commented on xtra-libp2p-ping/src/protocol.rs:
```diff
@@ -0,0 +1,46 @@
+}
+
+/// Waits for a ping and sends a pong.
+pub(crate) async fn recv(mut stream: Substream) -> io::Result<()> {
```
❓ What is the point of separating these functions into `protocol`?

I know that the initial design also did that, but I have still not figure out what the advantage would be. I think it is way easier to understand if we inlined this code in the respective actor. 
For me the actor represents the protocol - there is not really a way (at least I don't see a clear one yet...) to separate "the protocol" (as in the pure message exchange) from the actor. With ping now using two actors it makes even less sense to me to have these functions separate from the actor. I feel it just makes it harder to follow the code.

luckysori replied:
The ping protocol we are performing doesn't have anything to do with the substream that we send messages through.

I've introduced 18dbe20e8c12fcd09fb4743518c5c872ffd740b5 to demonstrate how this doesn't need to depend on `xtra`.

The message protocol is on a different level of abstraction than the actors that call its functions. For example, the `ping` actor calculates latencies based on performing the protocol. The message protocol could also be tested in isolation.

da-kami replied:
It will be interesting to "map" this information to other protocols we design. For rollover we somehow have a similar constraint:
The `setup_contract::roll_over` function is the "protocol" and everything else is only scaffolding to *start* that protocol - including the maker's accept and reject. 

But, when it comes to introducing a new version of the protocol then that would be at level of negotiating the substream - and that is done by the actor. So, to have a new protocol version we would need another actor as well, which renders this separation somewhat useless. Or am I missing something?

da-kami reviewed (approved):
One comment on how to structure the code of these protocols. I was thinking about this when implementing rollover as well. Looking forward to discussing some of the findings here in context of the rollover impl :)

luckysori committed:
Make xtra_libp2p_ping::protocol generic over Async{Write,Read}Ext

This highlights that the ping protocol defined here does not need
`xtra` or `xtra_libp2p` to be used.
